### PR TITLE
Create restart()

### LINKED ISSUE #1637: Suggestion: restart() function
Currently, there's no automatic way to restart a chip, it always has to be done manually via toolgun.

Having something like E2's `restart()` on serverside would be handy, even more so if you could optionally target a different chip owned by yourself.

### PR DESCRIPTION
**Pull request by legokidlogan**
Resolves #1637

### CODE REVIEW

File: lua/starfall/libs_sh/builtins.lua [hunk 1]
```diff
@@ -596,6 +597,35 @@ if SERVER then
...
+		if (chip.nextRestartTime or 0) > now then SF.Throw("That starfall is on restart() cooldown", 2) end
+
+		chip.nextRestartTime = now + restartCooldown:GetFloat()
+
+		timer.Simple(0, function()
+			chip = inst.entity
+			if not IsValid(chip) then return end -- Could happen during entity removal
+
+			chip:SetupFiles(chip.sfdata)
+		end)
```
Comment (thegrb93) [inline lua/starfall/libs_sh/builtins.lua:627]:
```suggestion
		if chip then
			chip = getent(chip)
			if not (chip.Starfall and chip.instance) then SF.Throw("Entity has no starfall instance", 2) end
			if chip.owner ~= instance.player then SF.Throw("You don't own that starfall", 2) end
		else
			chip = instance.entity
		end

		local now = CurTime()
		if (chip.nextRestartTime or 0) > now then SF.Throw("That starfall is on restart() cooldown", 2) end

		chip.nextRestartTime = now + restartCooldown:GetFloat()

		timer.Simple(0, function()
			if IsValid(chip) then chip:SetupFiles(chip.sfdata) end
		end)
```

**Review by thegrb93 — APPROVED** [review]

**Comment by thegrb93** [comment]
Could the function work in clientside too?

**Comment by legokidlogan** [comment]
Yeah, I should be able to do that. [Clientside restart via context menu](https://github.com/thegrb93/StarfallEx/blob/master/lua/entities/starfall_processor/shared.lua#L161-L163) uses `chip:Compile()`, so I'll go with that.

File: lua/starfall/libs_sh/builtins.lua [hunk 2]
```diff
@@ -1191,6 +1193,34 @@ function builtins_library.enableHud(ply, active)
...
+
+	chip.nextRestartTime = now + restartCooldown:GetFloat()
+
+	timer.Simple(0, function()
+		if IsValid(chip) then
+			if SERVER then
+				chip:SetupFiles(chip.sfdata)
+			else
+				chip:Compile()
+			end
```
Comment (thegrb93) [inline lua/starfall/libs_sh/builtins.lua:1219]: I think chip:Compile() should work for both

**Comment by thegrb93** [comment]
I don't like the idea of having access to the ability to send files, even if its on a 5s delay. The toolgun has logic that prevents sending more than one chip at a time, but this doesn't, so someone can make a really big chip and spam sending it. We should only give access to 'Compile'.

**Review by thegrb93 — APPROVED** [review]

File: lua/starfall/libs_sh/builtins.lua [hunk 3]
```diff
@@ -1191,6 +1193,30 @@ function builtins_library.enableHud(ply, active)
...
+		if not (chip.Starfall and chip.instance) then SF.Throw("Entity has no starfall instance", 2) end
+		if chip.owner ~= instance.player then SF.Throw("You don't own that starfall", 2) end
+	else
+		chip = instance.entity
+	end
+
+	local now = CurTime()
+	if (chip.nextRestartTime or 0) > now then SF.Throw("That starfall is on restart() cooldown", 2) end
+
+	chip.nextRestartTime = now + restartCooldown:GetFloat()
```
Comment (thegrb93) [inline lua/starfall/libs_sh/builtins.lua:1211]: One last thing, restartCooldown should be clamped with a minimum time. Maybe 0.1 second.

**Pull request merged**